File: internal/fetch/fetch.go
// Package fetch provides an HTTP fetcher for retrieving manifest bytes.
//
// The Fetcher interface is intentionally minimal so that callers can inject
// fakes in tests without touching the network.
package fetch

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"time"
)

// ErrHTTPStatus is returned when a fetch receives a non-2xx HTTP response.
type ErrHTTPStatus struct {
	URL    string
	Status int
}

func (e *ErrHTTPStatus) Error() string {
	return fmt.Sprintf("fetch %s: unexpected status %d", e.URL, e.Status)
}

// Fetcher retrieves the raw bytes located at a URL.
type Fetcher interface {
	Fetch(ctx context.Context, url string) ([]byte, error)
}

// httpFetcher is the default Fetcher backed by net/http.
type httpFetcher struct {
	client *http.Client
	// maxBytes caps the response size to avoid unbounded memory use.
	maxBytes int64
}

var _ Fetcher = (*httpFetcher)(nil)

// Option configures the httpFetcher.
type Option func(*httpFetcher)

// WithClient sets a custom *http.Client.
func WithClient(c *http.Client) Option {
	return func(f *httpFetcher) { f.client = c }
}

// WithMaxBytes caps the number of bytes read from a response body.
func WithMaxBytes(n int64) Option {
	return func(f *httpFetcher) { f.maxBytes = n }
}

// New returns the default HTTP-backed Fetcher.
func New(opts ...Option) Fetcher {
	f := &httpFetcher{
		client:   &http.Client{Timeout: 30 * time.Second},
		maxBytes: 64 << 20, // 64 MiB
	}
	for _, opt := range opts {
		opt(f)
	}
	return f
}

// Fetch performs a GET request and returns the response body.
func (f *httpFetcher) Fetch(ctx context.Context, url string) ([]byte, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, fmt.Errorf("build request for %s: %w", url, err)
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "storyquery")

	resp, err := f.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("get %s: %w", url, err)
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, &ErrHTTPStatus{URL: url, Status: resp.StatusCode}
	}

	body, err := io.ReadAll(io.LimitReader(resp.Body, f.maxBytes))
	if err != nil {
		return nil, fmt.Errorf("read body from %s: %w", url, err)
	}
	return body, nil
}

File: src/fetch.ts
// Retrieves raw manifest text over HTTP using the native fetch API.
//
// The Fetcher type is intentionally minimal so callers can inject fakes in tests
// without touching the network.

/** Thrown when a fetch receives a non-2xx HTTP response. Drives exit code 3. */
export class HttpStatusError extends Error {
  readonly url: string;
  readonly status: number;
  constructor(url: string, status: number) {
    super(`fetch ${url}: unexpected status ${status}`);
    this.name = "HttpStatusError";
    this.url = url;
    this.status = status;
  }
}

/** Retrieves the raw text located at a URL. */
export type Fetcher = (url: string, signal?: AbortSignal) => Promise<string>;

const DEFAULT_TIMEOUT_MS = 30_000;
const DEFAULT_MAX_BYTES = 64 << 20; // 64 MiB

export interface FetchOptions {
  timeoutMs?: number;
  maxBytes?: number;
}

/** Builds the default fetch-backed Fetcher. */
export function createFetcher(opts: FetchOptions = {}): Fetcher {
  const timeoutMs = opts.timeoutMs ?? DEFAULT_TIMEOUT_MS;
  const maxBytes = opts.maxBytes ?? DEFAULT_MAX_BYTES;

  return async (url: string, signal?: AbortSignal): Promise<string> => {
    const ctrl = new AbortController();
    const onAbort = () => ctrl.abort(signal?.reason);
    if (signal) {
      if (signal.aborted) ctrl.abort(signal.reason);
      else signal.addEventListener("abort", onAbort, { once: true });
    }
    const timer = setTimeout(() => ctrl.abort(new Error("request timed out")), timeoutMs);

    try {
      const resp = await fetch(url, {
        method: "GET",
        headers: { Accept: "application/json", "User-Agent": "storyquery" },
        signal: ctrl.signal,
      });

      if (resp.status < 200 || resp.status >= 300) {
        throw new HttpStatusError(url, resp.status);
      }
      return await readCapped(resp, url, maxBytes);
    } finally {
      clearTimeout(timer);
      if (signal) signal.removeEventListener("abort", onAbort);
    }
  };
}

/** Reads a response body as text, aborting if it exceeds maxBytes. */
async function readCapped(resp: Response, url: string, maxBytes: number): Promise<string> {
  if (!resp.body) return await resp.text();

  const reader = resp.body.getReader();
  const chunks: Uint8Array[] = [];
  let total = 0;
  try {
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      if (value) {
        total += value.byteLength;
        if (total > maxBytes) {
          throw new Error(`read body from ${url}: response exceeds ${maxBytes} bytes`);
        }
        chunks.push(value);
      }
    }
  } finally {
    reader.releaseLock();
  }

  return Buffer.concat(chunks).toString("utf8");
}
